### ISSUE TITLE
Deployment fails with journald log-driver on predeploy

### ISSUE THREAD

**Issue by csenszike** [dokku/dokku#4376]
## Description of problem

I've an apps.json with predeploy command. With the latest release I cannot deploy my app with journald log-driver (with json-file driver it deploys normally):

```
root@agora:~# dokku ps:restart demo
-----> Releasing demo...
-----> Deploying demo...
-----> Checking for predeploy task
-----> Executing predeploy task from app.json: rails db:migrate
 !     Failed to create predeploy execution container: Error response from daemon: unknown log opt 'max-size' for journald log driver

2021/02/03 14:56:40 exit status 1
2021/02/03 14:56:40 exit status 1
2021/02/03 14:56:40 exit status 1
 !     exit status 1
```

### `dokku report APP_NAME` output

```
root@agora:~# dokku report demo
-----> uname: Linux agora 4.15.0-121-generic #123-Ubuntu SMP Mon Oct 5 16:16:40 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
-----> memory:
                     total        used        free      shared  buff/cache   available
       Mem:            985         320         170          18         494         489
       Swap:          2047          45        2002
-----> docker version:
       Client: Docker Engine - Community
        Version:           20.10.3
        API version:       1.41
        Go version:        go1.13.15
        Git commit:        48d30b5
        Built:             Fri Jan 29 14:33:13 2021
        OS/Arch:           linux/amd64
        Context:           default
        Experimental:      true

       Server: Docker Engine - Community
        Engine:
         Version:          20.10.3
         API version:      1.41 (minimum version 1.12)
         Go version:       go1.13.15
         Git commit:       46229ca
         Built:            Fri Jan 29 14:31:25 2021
         OS/Arch:          linux/amd64
         Experimental:     false
        containerd:
         Version:          1.4.3
         GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
        runc:
         Version:          1.0.0-rc92
         GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
        docker-init:
         Version:          0.19.0
         GitCommit:        de40ad0
-----> docker daemon info:
WARNING: No swap limit support
       Client:
        Context:    default
        Debug Mode: true
        Plugins:
         app: Docker App (Docker Inc., v0.9.1-beta3)
         buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)

       Server:
        Containers: 2
         Running: 2
         Paused: 0
         Stopped: 0
        Images: 16
        Server Version: 20.10.3
        Storage Driver: overlay2
         Backing Filesystem: extfs
         Supports d_type: true
         Native Overlay Diff: true
        Logging Driver: json-file
        Cgroup Driver: cgroupfs
        Cgroup Version: 1
        Plugins:
         Volume: local
         Network: bridge host ipvlan macvlan null overlay
         Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
        Swarm: inactive
        Runtimes: io.containerd.runtime.v1.linux runc io.containerd.runc.v2
        Default Runtime: runc
        Init Binary: docker-init
        containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
        runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
        init version: de40ad0
        Security Options:
         apparmor
         seccomp
          Profile: default
        Kernel Version: 4.15.0-121-generic
        Operating System: Ubuntu 18.04.5 LTS
        OSType: linux
        Architecture: x86_64
        CPUs: 1
        Total Memory: 985.2MiB
        Name: agora
        ID: WB7V:5TIJ:P2E5:7UNO:DM2F:OW76:DQCK:GNGB:JBXM:77W4:IH4P:BM6Q
        Docker Root Dir: /var/lib/docker
        Debug Mode: false
        Registry: https://index.docker.io/v1/
        Labels:
        Experimental: false
        Insecure Registries:
         127.0.0.0/8
        Live Restore Enabled: false

-----> git version: git version 2.17.1
-----> sigil version: 0.6.0
-----> herokuish version:
       herokuish: 0.5.24
       buildpacks:
         heroku-buildpack-multi     v1.0.0
         heroku-buildpack-ruby      v222
         heroku-buildpack-nodejs    v182
         heroku-buildpack-clojure   v86
         heroku-buildpack-python    v188
         heroku-buildpack-java      v69
         heroku-buildpack-gradle    v34
         heroku-buildpack-scala     v88
         heroku-buildpack-play      v26
         heroku-buildpack-php       v188
         heroku-buildpack-go        v149
         buildpack-nginx            v14
-----> dokku version: dokku version 0.23.1
-----> plugn version: plugn: 0.6.1
-----> dokku plugins:
         00_dokku-standard    0.23.1 enabled    dokku core standard plugin
         20_events            0.23.1 enabled    dokku core events logging plugin
         app-json             0.23.1 enabled    dokku core app-json plugin
         apps                 0.23.1 enabled    dokku core apps plugin
         builder-cnb          0.23.1 enabled    dokku core builder-cnb plugin
         builder-dockerfile   0.23.1 enabled    dokku core builder-dockerfile plugin
         builder-herokuish    0.23.1 enabled    dokku core builder-herokuish plugin
         buildpacks           0.23.1 enabled    dokku core buildpacks plugin
         certs                0.23.1 enabled    dokku core certificate management plugin
         checks               0.23.1 enabled    dokku core checks plugin
         common               0.23.1 enabled    dokku core common plugin
         config               0.23.1 enabled    dokku core config plugin
         cron                 0.23.1 enabled    dokku core cron plugin
         docker-options       0.23.1 enabled    dokku core docker-options plugin
         domains              0.23.1 enabled    dokku core domains plugin
         enter                0.23.1 enabled    dokku core enter plugin
         git                  0.23.1 enabled    dokku core git plugin
         letsencrypt          0.9.4 enabled    Automated installation of let's encrypt TLS certificates
         logs                 0.23.1 enabled    dokku core logs plugin
         network              0.23.1 enabled    dokku core network plugin
         nginx-vhosts         0.23.1 enabled    dokku core nginx-vhosts plugin
         plugin               0.23.1 enabled    dokku core plugin plugin
         postgres             1.11.6 enabled    dokku postgres service plugin
         proxy                0.23.1 enabled    dokku core proxy plugin
         ps                   0.23.1 enabled    dokku core ps plugin
         repo                 0.23.1 enabled    dokku core repo plugin
         resource             0.23.1 enabled    dokku core resource plugin
         scheduler-docker-local 0.23.1 enabled    dokku core scheduler-docker-local plugin
         shell                0.23.1 enabled    dokku core shell plugin
         ssh-keys             0.23.1 enabled    dokku core ssh-keys plugin
         storage              0.23.1 enabled    dokku core storage plugin
         tags                 0.23.1 enabled    dokku core tags plugin
         tar                  0.23.1 enabled    dokku core tar plugin
         trace                0.23.1 enabled    dokku core trace plugin
=====> demo app information
       App deploy source:
       App dir:                       /home/dokku/demo
       App locked:                    false
=====> demo buildpacks information
       Buildpacks computed stack:     gliderlabs/herokuish:latest
       Buildpacks global stack:
       Buildpacks list:
       Buildpacks stack:
xargs: unmatched single quote; by default quotes are special to xargs unless you use the -0 option
CN = demo.katadmin.hu
error 20 at 0 depth lookup: unable to get local issuer certificate
=====> demo ssl information
       Ssl dir:                       /home/dokku/demo/tls
       Ssl enabled:                   true
       Ssl hostnames:                 demo.katadmin.hu www.demo.katadmin.hu
       Ssl expires at:                May  4 10:38:54 2021 GMT
       Ssl issuer:                    C = US, O =
       Ssl starts at:                 Feb  3 10:38:54 2021 GMT
       Ssl subject:                   subject=CN = demo.katadmin.hu
       Ssl verified:                  self signed
=====> demo checks information
       Checks disabled list:          none
       Checks skipped list:           none
=====> demo cron information
       Cron task count:               0
=====> demo docker options information
       Docker options build:          --link dokku.postgres.demo:dokku-postgres-demo --log-driver=journald --log-opt tag=demo
       Docker options deploy:         --link dokku.postgres.demo:dokku-postgres-demo --log-driver=journald --log-opt tag=demo --restart=on-failure:10 -v /www/demo:/app/public/uploads
       Docker options run:            --link dokku.postgres.demo:dokku-postgres-demo --log-driver=journald --log-opt tag=demo -v /www/demo:/app/public/uploads
=====> demo domains information
       Domains app enabled:           true
       Domains app vhosts:            demo.katadmin.hu www.demo.katadmin.hu
       Domains global enabled:        true
       Domains global vhosts:         katadmin.hu
=====> demo git information
       Git deploy branch:             master
       Git global deploy branch:      master
       Git keep git dir:              false
       Git rev env var:               GIT_REV
       Git sha:                       198f66a
       Git last updated at:           1612352061
=====> demo logs information
       Logs computed max size:        10m
       Logs global max size:          10m
       Logs global vector sink:
       Logs max size:
       Logs vector sink:
=====> demo network information
       Network attach post create:
       Network attach post deploy:
       Network bind all interfaces:   false
       Network web listeners:         172.17.0.4:5000
=====> demo nginx information
       Nginx access log format:
       Nginx access log path:         /var/log/nginx/demo-access.log
       Nginx bind address ipv4:
       Nginx bind address ipv6:       ::
       Nginx client max body size:
       Nginx disable custom config:   false
       Nginx error log path:          /var/log/nginx/demo-error.log
       Nginx global hsts:             true
       Nginx computed hsts:           true
       Nginx hsts:
       Nginx hsts include subdomains: true
       Nginx hsts max age:            15724800
       Nginx hsts preload:            false
       Nginx proxy buffer size:       4096
       Nginx proxy buffering:         on
       Nginx proxy buffers:           8 4096
       Nginx proxy busy buffers size: 8192
       Nginx proxy read timeout:      60s
       Nginx last visited at:         1612360271
       Nginx x forwarded for value:   $remote_addr
       Nginx x forwarded port value:  $server_port
       Nginx x forwarded proto value: $scheme
=====> demo proxy information
       Proxy enabled:                 true
       Proxy port map:                http:80:5000 https:443:5000
       Proxy type:                    nginx
=====> demo ps information
       Deployed:                      true
       Processes:                     1
       Ps can scale:                  true
       Ps restart policy:             on-failure:10
       Restore:                       true
       Running:                       true
       Status web 1:                  running (CID: 7722d53cdf3)
=====> demo resource information
=====> demo scheduler-docker-local information
       Scheduler docker local disable chown:
=====> demo storage information
       Storage build mounts:
       Storage deploy mounts:         -v /www/demo:/app/public/uploads
       Storage run mounts:            -v /www/demo:/app/public/uploads

```

**Comment by josegonzalez** [comment]
I wonder if we can parse the incoming options and drop the max-size if we detect anything other than `json-file`. I think thats probably acceptable.

Would probably break if the user sets it at the daemon level though...

**Comment by csenszike** [comment]
It seems the bug is because of default log max-size.
If I set it globally to unlimited (found instructions in the docs), the deployment runs normally:

```
root@agora:~# dokku logs:report
=====> demo logs information
       Logs computed max size:        10m
       Logs global max size:          10m
       Logs global vector sink:
       Logs max size:
       Logs vector sink:
root@agora:~# dokku logs:set --global max-size unlimited
=====> Setting max-size to unlimited
root@agora:~# dokku logs:report
=====> demo logs information
       Logs computed max size:        unlimited
       Logs global max size:          unlimited
       Logs global vector sink:
       Logs max size:
       Logs vector sink:
```

Of course it's a dirt fix, but maybe a good starting point. The main question in my opinion is what causes the bug in the latest release I've never faced any problems with journald driver earlier (even didn't know about these global log options).

**Comment by josegonzalez** [comment]
Thats a new thing we added in the latest version (0.23.0). I didn't realize that would break installs for people, but I'll add it to the migration guide.

That said, I think maybe we can be a bit nicer about how we handle this.

**Comment by josegonzalez** [comment]
Closing as there is a pull request open.

**Comment by madwort** [comment]
FWIW I'm still experiencing this bug in`0.23.9` and `0.24.x` (fix should be in `0.23.2` upwards, right?) when upgrading on debian stretch from `0.22.x` on a server currently using `journald` for logs

**Comment by josegonzalez** [comment]
@madwort mind filing a new issue to track the problem? I'm guessing maybe something else is going on, but having all the debug info would be great.
